File: server-pedidos-java/src/main/java/br/edu/utfpr/serverpedidosjava/controller/dto/PedidoProdutoDTO.java
package br.edu.utfpr.serverpedidosjava.controller.dto;

import java.math.BigDecimal;

import br.edu.utfpr.serverpedidosjava.entity.Pedido;
import br.edu.utfpr.serverpedidosjava.entity.PedidoProduto;
import br.edu.utfpr.serverpedidosjava.entity.PedidoProdutoId;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Positive;

public record PedidoProdutoDTO(
        @NotNull(message = "{produto.notnull}") ProdutoDTO produto,
        @NotNull(message = "{valorunitario.notnull}") @Positive(message = "{valorunitario.positive}") BigDecimal valorUnitario,
        @NotNull(message = "{quantidade.notnull}") @Positive(message = "{quantidade.positive}") BigDecimal quantidade) {

    public PedidoProduto toEntity(Pedido pedido) {
        PedidoProdutoId pedidoProdutoId = new PedidoProdutoId();
        pedidoProdutoId.setPedido(pedido);
        pedidoProdutoId.setProduto(this.produto.toEntity());

        PedidoProduto pedidoProduto = new PedidoProduto();
        pedidoProduto.setId(pedidoProdutoId);
        pedidoProduto.setValorUnitario(this.valorUnitario);
        pedidoProduto.setQuantidade(this.quantidade);

        return pedidoProduto;
    }

    public static PedidoProdutoDTO fromEntity(PedidoProduto pedidoProduto) {
        return new PedidoProdutoDTO(
                ProdutoDTO.fromEntity(pedidoProduto.getId().getProduto()),
                pedidoProduto.getValorUnitario(),
                pedidoProduto.getQuantidade());
    }
}

File: server-pedidos/src/main/kotlin/br/edu/utfpr/serverpedidos/controller/dto/PedidoProdutoDTO.kt
package br.edu.utfpr.serverpedidos.controller.dto

import br.edu.utfpr.serverpedidos.entity.Pedido
import br.edu.utfpr.serverpedidos.entity.PedidoProduto
import br.edu.utfpr.serverpedidos.entity.PedidoProdutoId
import jakarta.validation.constraints.NotNull
import jakarta.validation.constraints.Positive
import java.math.BigDecimal

data class PedidoProdutoDTO(
    @field:NotNull(message = "{produto.notnull}")
    val produto: ProdutoDTO? = null,
    @field:Positive(message = "{valorunitario.positive}")
    val valorUnitario: BigDecimal = BigDecimal.ZERO,
    @field:Positive(message = "{quantidade.positive}")
    val quantidade: BigDecimal = BigDecimal.ZERO
) {
    fun toEntity(pedido: Pedido): PedidoProduto = PedidoProduto(
        id = PedidoProdutoId(
            pedido = pedido,
            produto = this.produto?.toEntity()
        ),
        valorUnitario = this.valorUnitario,
        quantidade = this.quantidade
    )

    companion object {
        fun fromEntity(pedidoProduto: PedidoProduto): PedidoProdutoDTO = PedidoProdutoDTO(
            produto = pedidoProduto.id.produto?.let { ProdutoDTO.fromEntity(it) },
            valorUnitario = pedidoProduto.valorUnitario,
            quantidade = pedidoProduto.quantidade
        )
    }
}
